controllers/remote/cluster_cache.go

```diff
@@ -0,0 +1,203 @@
+		For(&clusterv1.Cluster{}).
+		WithOptions(input.ControllerOptions).
+		Build(m)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create cluster cache manager controller")
+	}
+
+	return m, nil
+}
+
+// ClusterCache returns the ClusterCache for cluster, creating a new ClusterCache if needed.
+func (m *ClusterCacheManager) ClusterCache(ctx context.Context, cluster client.ObjectKey) (ClusterCache, error) {
+	cache := m.getClusterCache(cluster)
+	if cache != nil {
+		return cache, nil
+	}
+
+	return m.newClusterCache(ctx, cluster)
+}
+
+// getClusterCache returns the ClusterCache for cluster, or nil if it does not exist.
+func (m *ClusterCacheManager) getClusterCache(cluster client.ObjectKey) ClusterCache {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	return m.clusterCaches[cluster]
+}
+
+// defaultNewClusterCache creates and starts a new ClusterCache for cluster.
+func (m *ClusterCacheManager) newClusterCache(ctx context.Context, cluster client.ObjectKey) (ClusterCache, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// If another goroutine created the cache while this one was waiting to acquire the write lock, return that
+	// instead of overwriting it.
+	if c, exists := m.clusterCaches[cluster]; exists {
+		return c, nil
+	}
+
+	config, err := m.newRESTConfig(ctx, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating REST client config for remote cluster")
+	}
+
+	remoteCache, err := cache.New(config, cache.Options{})
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating cache for remote cluster")
+	}
+	stop := make(chan struct{})
+
+	cc := &clusterCache{
+		Cache: remoteCache,
+		stop:  stop,
+	}
+	m.clusterCaches[cluster] = cc
+
+	// Start the cache!!!
+	go remoteCache.Start(cc.stop)
+
+	return cc, nil
+}
+
+func (m *ClusterCacheManager) defaultNewRESTConfig(ctx context.Context, cluster client.ObjectKey) (*rest.Config, error) {
+	config, err := RESTConfig(ctx, m.managementClusterClient, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching remote cluster config")
+	}
+	return config, nil
+}
+
+// Reconcile reconciles Clusters and removes ClusterCaches for any Cluster that cannot be retrieved from the
+// management cluster.
+func (m *ClusterCacheManager) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	ctx := context.Background()
+
+	log := m.log.WithValues("namespace", req.Namespace, "name", req.Name)
+	log.V(4).Info("Reconciling")
+
+	var cluster clusterv1.Cluster
+
+	err := m.managementClusterClient.Get(ctx, req.NamespacedName, &cluster)
+	if err == nil {
+		log.V(4).Info("Cluster still exists")
+		return reconcile.Result{}, nil
+	}
+
+	log.V(4).Info("Error retrieving cluster", "error", err.Error())
+
+	c := m.getClusterCache(req.NamespacedName)
+	if c == nil {
+		log.V(4).Info("No current cluster cache exists - nothing to do")
+	}
+
+	log.V(4).Info("Stopping cluster cache")
+	c.Stop()
+
+	delete(m.clusterCaches, req.NamespacedName)
```

Comment:
Should there be a method on `ClusterCacheManager` that wraps this with a lock? It seems a bit weird that we are locking on insertion and reading but not deletion.

Reply:
Yeah I realized this last night. I have it fixed locally and it'll be in the next push.